thor/src/protocols/splice.rs

```diff
@@ -147,31 +153,47 @@ impl State0 {
         Message0 {
             R: R_other,
             Y: Y_other,
-            splice_in: splice_in_other,
+            splice: splice_other,
         }: Message0,
     ) -> Result<State1> {
-        let previous_funding_txin = self.previous_tx_f.as_txin();
-        let previous_funding_psbt = PartiallySignedTransaction::from_unsigned_tx(Transaction {
-            version: 2,
-            lock_time: 0,
-            input: vec![previous_funding_txin],
-            output: vec![],
-        })
-        .expect("Only fails if script_sig or witness is empty which is not the case.");
-
         let mut our_balance = self.previous_balance.ours;
         let mut their_balance = self.previous_balance.theirs;
 
+        let mut splice_outputs = vec![];
         let mut splice_in_inputs = vec![];
 
-        if let Some(splice_in) = self.splice_in_self.clone() {
-            splice_in_inputs.push(splice_in.input_psbt);
-            our_balance += splice_in.amount;
+        match self.splice_self.clone() {
+            Splice::Out(tx_out) => {
+                if tx_out.value > self.previous_balance.ours.as_sat() {
+                    anyhow::bail!("We are splicing out more than we have");
+                } else {
+                    our_balance -= Amount::from_sat(tx_out.value) + Amount::from_sat(TX_FEE);
+                    splice_outputs.push(tx_out);
+                }
+            }
+            Splice::In { amount, input_psbt } => {
+                splice_in_inputs.push(input_psbt);
+                our_balance += amount;
+            }
+            Splice::None => (),
         }
 
-        if let Some(splice_in) = splice_in_other {
-            splice_in_inputs.push(splice_in.input_psbt);
-            their_balance += splice_in.amount;
+        match splice_other {
+            Splice::Out(tx_out) => {
+                if tx_out.value > self.previous_balance.theirs.as_sat() {
+                    anyhow::bail!("Counterpart is splicing out more than they have");
+                } else {
+                    // Need to pay the transaction fee, taking it out of the splice out.
+                    // TODO: split between splice in and splice out if there is both
+                    their_balance -= Amount::from_sat(tx_out.value) + Amount::from_sat(TX_FEE);
+                    splice_outputs.push(tx_out);
+                }
+            }
+            Splice::In { amount, input_psbt } => {
+                splice_in_inputs.push(input_psbt);
+                their_balance += amount;
+            }
+            Splice::None => (),
         }
 
         // Sort the PSBT inputs based on the ascending lexicographical order of
```

Comment:
:question: I pointed out in the last PR that it would be nicer if we didn't require callers of `SpliceTransaction::new` to pre-sort the inputs, because if the sorting is only enforced in one place (inside the constructor for each transaction type) then it's easier to document and more consistent.

You said it was hard to do, but I don't understand why without more context. AFAIK you don't need to rely on the order of the inputs outside of the `SpliceTransaction` constructor: you should be able to sign the input coming from the previous TX_f manually, and let the user wallets sign the new splice-in inputs (if any).